service/src/main/java/bio/terra/tanagra/plugin/PluginService.java

```diff
@@ -0,0 +1,74 @@
+package bio.terra.tanagra.plugin;
+
+import bio.terra.tanagra.service.UnderlaysService;
+import bio.terra.tanagra.underlay.Underlay;
+import java.util.HashMap;
+import java.util.Map;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+@Component
+public class PluginService {
+  private final UnderlaysService underlayService;
+  private final Map<String, Plugin> loadedPlugins = new HashMap<>();
+
+  @Autowired
+  public PluginService(UnderlaysService underlayService) throws PluginException {
+    this.underlayService = underlayService;
+  }
+
+  public <T extends Plugin> T getPlugin(String underlayName, Class<T> c) {
```

Comment:
I did it this way because when getting when getting a plugin the developer doesn't actually care about the type, they care about the interface. If this were to take in the `PluginType` enum, the developer would then also have to know what type to cast the returned plugin to. As implemented, the developer only needs to specify the interface they care about, and they get back that interface.

Reply:
That ignores the advantages of generics and requires more boilerplate code be added in more places for each new plugin we create. Are generics disallowed in Tanagra?

Reply:
But the caller does care that it's an implementation of `AccessControlPlugin`? In order to access any members of the plugin the returned class needs to be of that type.

To be clear, what I'm suggesting here is:
```java
WhateverPlugin wp = pluginService.getPlugin(Interface.class, underlayName);
```

_not:_
```java
WhateverPlugin wp = pluginService.getPlugin(Implementation.class, underlayName);
```

If we get plugins using the enum, then the call to get a plugin would be (which has an unnecessary cast which could be handed elsewhere:
```java
WhateverPlugin wp = (WhateverPlugin)pluginService.getPlugin(PluginType.WHATEVER, underlayName);
```

Regarding adding a `getWhateverPlugin()` method to the plugin service for _every_ plugin... What I don't understand is why this is preferred. It is essentially replicating what the Java compiler is doing anyways via type erasure, is another boilerplate step that needs to be copy/pasted for each new plugin, and is more code that needs to be maintained.

Reply:
In your first and last code snippets, wouldn't it be `WhateverInterface` not `WhateverPlugin` as the variable type?

I think we're actually getting at the same thing -- the `Class<T> c` method argument is confusing and unnecessarily vague. It's specified as just a class name (no restrictions), when what we expect is that it's one of the (currently 2) interfaces that extend `Plugin`: `AccessControlPlugin` or `IdentityPlugin`. So I think you should change this method to make that clear and impossible for callers to mess up. Using the enum is one way -- agreed the casting required makes that less desirable. Defining two methods each with the correct return type, is another way. Or please propose something else too, I'm not wedded to these ideas.

RE adding new plugins: I think we're mixing up adding new plugin types and adding new plugin implementations. Adding a new plugin implementation for e.g. access control, wouldn't require any changes here. Adding a new plugin type would, but that's a much more invasive change and would require lots of other changes also.

Reply:
My examples were confusing because Tanagra doesn't use `Ixxx` or `xxxInterface` naming for plugins, so `NamePlugin` is an interface, and then I was trying to differentiate between a the interface and the implementation in the method arguments and just further confused the issue. My bad.

I'm not sure what you mean that there are no restrictions on the `Class<T>` parameter of `getPlugin()`, it is type restricted to classes inheriting from `Plugin` via `<T extends Plugin>` and won't compile if anything else is passed to it.

I wasn't referring to implementations when talking about adding new plugins, I was talking about new plugin types. It seems like adding a `getAccessControlPlugin()`, `getIdentityPlugin()`, and `getWhateverPlugin()` for each plugin type is redundant as this functionality is already exposed by `getPlugin()`.

Reply:
Yes, the class name is restricted to `T extends Plugin` but that means passing an implementation class name would be valid right, though I think that's not what you intended? It'd be great if there were a way to say `T extends Plugin but only immediate sub-interfaces no implementation classes`, but I don't know how to do that with Java type parameters. That's why I suggested just breaking it out into separate methods, but if you can figure out another way to make sure callers don't pass implementation class names at compile-time, I'm on board with that.